### PR TITLE
Split Poll into Poll and Register

### LINKED ISSUE #757: Split `Poll` into `Poll` and `Register`.
As part of [rewriting the readiness queue](https://github.com/carllerche/mio/commit/43a69f218957ea7f43bb821734dbfbad09078c6b), `Poll::poll` acquired a lock. This is because `Poll` needed to be `Sync` so that I/O resources could be merged from other threads.

However, `Poll` really provides two separate capabilities:

* Registering I/O resources.
* Polling for I/O events.

Only I/O registration needs to be `Sync`. If the `Poll` type is split into two types, then `Poll::poll` can take `&mut self` and the lock can be removed.

** But, what about polling from multiple threads? **

Well, that isn't supported today anyway. I also suspect that the answer isn't to make `Poll` `Sync` but to add some clone type functionality so that each thread can get its own `Poll` handle. This would also better support the max concurrency settings that IOCP has.

### PR DESCRIPTION
**Pull request by carllerche**
This also makes `Poll::poll` require `&mut self` since `Poll` is no
longer expected to be stored in a `Sync` context.

Resolves #757

This is a very large change and I would not look for perfection, just that nothing in the change is terribly broken. This will be but one of many PRs to transition Mio to 0.7 and I am attempting to make it as incremental as possible.

cc @Thomasdezeeuw @stjepang

### CODE REVIEW

File: src/poll.rs [hunk 1]
```diff
@@ -652,25 +654,233 @@ impl Poll {
...
+
+impl fmt::Debug for Poll {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.debug_struct("Poll").finish()
+    }
+}
+
+impl fmt::Debug for Register {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.debug_struct("Poll").finish()
```
Comment (Thomasdezeeuw) [inline src/poll.rs]: This uses `Poll`, but it's for `Register`.

File: test/test_custom_evented.rs [hunk 2]
```diff
@@ -176,120 +197,6 @@ mod stress {
         }
     }
 
-    #[test]
-    fn multi_threaded_poll() {
```
Comment (Thomasdezeeuw) [inline test/test_custom_evented.rs]: Is this test coming back?

Comment (carllerche) [reply test/test_custom_evented.rs]: No since we are getting rid of multi threaded polling (removal of the locks).

**Comment by carllerche** [comment]
@Thomasdezeeuw I agree with you about the tests and `poll.register().register(...)`. However, I hope to punt that until later.

- Many tests will be deleted (as we delete old functionality).
- Tests need a general overhaul. I think this would be best done after the Mio API changes are done.
- I am not sure exactly sure what the best strategy will be for us. We may just want to bring back `Poll::register`, `reregister`, `deregister` and rename `register()` -> `handle()` or something. I think we can better evaluate that after the changes we know we want are done.

Thoughts?

**Comment by carllerche** [comment]
@seanmonstar @Thomasdezeeuw I don't have a huge opinion re: what `Register` should be called. If we want to change to `Registry`, we can do that.

**Comment by hawkw** [comment]
FWIW, I'm +1 on renaming `Register` to `Registry`

**Comment by stjepang** [comment]
Same here. Even though `Register` can be a noun, it sounds more like an action to me, while `Registry` would be unambiguous.

**Comment by Thomasdezeeuw** [comment]
`Registry` is fine by me.

**Pull request merged**